cipher_suite.go

```diff
@@ -85,3 +87,45 @@ func encodeCipherSuites(c []cipherSuite) []byte {
 
 	return out
 }
+
+func parseCipherSuites(userSelectedSuites []CipherSuiteID, psk []byte) ([]cipherSuite, error) {
+	cipherSuitesForIDs := func(ids []CipherSuiteID) ([]cipherSuite, error) {
+		cipherSuites := []cipherSuite{}
+		for _, id := range ids {
+			c := cipherSuiteForID(id)
+			if c == nil {
+				return nil, fmt.Errorf("CipherSuite with id(%d) is not valid", id)
+			}
+			cipherSuites = append(cipherSuites, c)
+		}
+		return cipherSuites, nil
+	}
+
+	var (
+		cipherSuites []cipherSuite
+		err          error
+		i            int
+	)
+	if len(userSelectedSuites) != 0 {
+		cipherSuites, err = cipherSuitesForIDs(userSelectedSuites)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		cipherSuites = defaultCipherSuites()
+	}
+
+	for _, c := range cipherSuites {
+		if (psk != nil && c.isPSK()) || (psk == nil && !c.isPSK()) {
+			cipherSuites[i] = c
+			i++
+		}
+	}
```

Comment:
Wouldn't this loop potentially result in cipher suites with and without PSK being considered valid? As in you could end up with a mix of PSK and server certificate. I'm not sure if that's an actual problem as long as both the PSK and the certs are provided but I don't think I've ever seen that in the wild.

Reply:
Since `psk` doesn't change during this loop, I don't think this will be the case?
Only PSK suites will be selected if `psk != nil` and vice-versa.

Reply:
Updated! The intent of the function is to filter so you only get one type at the time (either PSK or non-PSK) 

I also assert earlier that you can't pass a certificate and a PSK.

I am trying to follow how OpenSSL acts here. If you pass a `psk` via the CLI they only do PSK suites.